pkg/services/authz/rbac/tree.go

```diff
@@ -0,0 +1,111 @@
+package rbac
+
+import (
+	"iter"
+
+	"github.com/grafana/grafana/pkg/services/authz/rbac/store"
+)
+
+func newFolderTree(folders []store.Folder) folderTree {
+	t := folderTree{
+		Index: make(map[string]int, len(folders)),
+		Nodes: make([]folderNode, 0, len(folders)),
+	}
+
+	for _, f := range folders {
+		t.Insert(f.UID, f.ParentUID)
+	}
+
+	return t
+}
+
+type folderTree struct {
+	// All nodes for the folderTree.
+	Nodes []folderNode
+	// Index is a map of folderNode UID to its positons in Nodes.
+	Index map[string]int
+}
+
+type folderNode struct {
+	// UID is the uniqiue identifier for folderNode
+	UID string
+	// Parent is the postion into folderTree nodes for parent, we store -1 for nodes that don't have a parent.
+	Parent int
+	// Children is positons into folderTree nodes for all children.
+	Children []int
+}
```

Comment:
We need to make all properties public so it can be encoded with json for caching